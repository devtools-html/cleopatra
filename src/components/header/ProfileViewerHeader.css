.profileViewerHeaderOverflowEdgeIndicator > .topEdge {
  top: 20px; /* Height of the time ruler */
}
