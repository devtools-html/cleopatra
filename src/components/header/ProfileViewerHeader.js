/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import React, { PureComponent } from 'react';
import ProfileThreadHeaderBar from './ProfileThreadHeaderBar';
import Reorderable from '../shared/Reorderable';
import TimeSelectionScrubber from './TimeSelectionScrubber';
import OverflowEdgeIndicator from './OverflowEdgeIndicator';
import explicitConnect from '../../utils/connect';
import {
  getProfile,
  getProfileViewOptions,
  getDisplayRange,
  getZeroAt,
} from '../../reducers/profile-view';
import { getHiddenThreads, getThreadOrder } from '../../reducers/url-state';

import {
  changeThreadOrder,
  updateProfileSelection,
  addRangeFilterAndUnsetSelection,
} from '../../actions/profile-view';

import type { Profile, ThreadIndex } from '../../types/profile';
import type { ProfileSelection } from '../../types/actions';
import type { Milliseconds, StartEndRange } from '../../types/units';
import type {
  ExplicitConnectOptions,
  ConnectedProps,
} from '../../utils/connect';

type StateProps = {|
  +profile: Profile,
  +selection: ProfileSelection,
  +threadOrder: ThreadIndex[],
  +hiddenThreads: ThreadIndex[],
  +timeRange: StartEndRange,
  +zeroAt: Milliseconds,
|};

type DispatchProps = {|
  +changeThreadOrder: typeof changeThreadOrder,
  +addRangeFilterAndUnsetSelection: typeof addRangeFilterAndUnsetSelection,
  +updateProfileSelection: typeof updateProfileSelection,
|};

type Props = ConnectedProps<{||}, StateProps, DispatchProps>;

class ProfileViewerHeader extends PureComponent<Props> {
  constructor(props: Props) {
    super(props);
    (this: any)._onZoomButtonClick = this._onZoomButtonClick.bind(this);
  }

  _onZoomButtonClick(start: Milliseconds, end: Milliseconds) {
    const { addRangeFilterAndUnsetSelection, zeroAt } = this.props;
    addRangeFilterAndUnsetSelection(start - zeroAt, end - zeroAt);
  }

  render() {
    const {
      profile,
      threadOrder,
      changeThreadOrder,
      selection,
      timeRange,
      zeroAt,
      hiddenThreads,
      updateProfileSelection,
    } = this.props;
    const threads = profile.threads;

    return (
      <TimeSelectionScrubber
        className="profileViewerHeader"
        zeroAt={zeroAt}
        rangeStart={timeRange.start}
        rangeEnd={timeRange.end}
        minSelectionStartWidth={profile.meta.interval}
        selection={selection}
        onSelectionChange={updateProfileSelection}
        onZoomButtonClick={this._onZoomButtonClick}
      >
        <OverflowEdgeIndicator className="profileViewerHeaderOverflowEdgeIndicator">
          {
            <Reorderable
              tagName="ol"
              className="profileViewerHeaderThreadList"
              order={threadOrder}
              orient="vertical"
              onChangeOrder={changeThreadOrder}
            >
              {threads.map((thread, threadIndex) =>
                <ProfileThreadHeaderBar
                  key={threadIndex}
                  threadIndex={threadIndex}
                  interval={profile.meta.interval}
                  rangeStart={timeRange.start}
                  rangeEnd={timeRange.end}
                  isHidden={hiddenThreads.includes(threadIndex)}
                  isModifyingSelection={selection.isModifying}
                />
              )}
            </Reorderable>
          }
        </OverflowEdgeIndicator>
      </TimeSelectionScrubber>
    );
  }
}

const options: ExplicitConnectOptions<{||}, StateProps, DispatchProps> = {
  mapStateToProps: state => ({
    profile: getProfile(state),
    selection: getProfileViewOptions(state).selection,
    threadOrder: getThreadOrder(state),
    hiddenThreads: getHiddenThreads(state),
    timeRange: getDisplayRange(state),
    zeroAt: getZeroAt(state),
  }),
  mapDispatchToProps: {
    changeThreadOrder,
    updateProfileSelection,
    addRangeFilterAndUnsetSelection,
  },
  component: ProfileViewerHeader,
};

export default explicitConnect(options);
