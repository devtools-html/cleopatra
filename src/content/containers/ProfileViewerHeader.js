/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import React, { PureComponent, PropTypes } from 'react';
import ProfileThreadHeaderBar from '../components/ProfileThreadHeaderBar';
import Reorderable from '../components/Reorderable';
import TimeSelectionScrubber from '../components/TimeSelectionScrubber';
import ProfileThreadJankOverview from './ProfileThreadJankOverview';
import ProfileThreadTracingMarkerOverview from './ProfileThreadTracingMarkerOverview';
import OverflowEdgeIndicator from '../components/OverflowEdgeIndicator';
import { connect } from 'react-redux';
import { getProfile, getProfileViewOptions, getThreadOrder, getDisplayRange, getZeroAt } from '../reducers/profile-view';
import actions from '../actions';

import type { Profile, ThreadIndex } from '../../common/types/profile';
import type { ProfileSelection } from '../actions/types';
import type { Milliseconds, StartEndRange } from '../../common/types/units';

type Props = {
  profile: Profile,
  className: string,
  threadOrder: ThreadIndex[],
  selection: ProfileSelection,
  timeRange: StartEndRange,
  zeroAt: Milliseconds,
  changeThreadOrder: ThreadIndex[] => void,
  updateProfileSelection: ProfileSelection => void,
  addRangeFilterAndUnsetSelection: (Milliseconds, Milliseconds) => void,
  changeSelectedThread: ThreadIndex => void,
};

class ProfileViewerHeader extends PureComponent {
  props: Props;

  constructor(props) {
    super(props);
    (this: any)._onZoomButtonClick = this._onZoomButtonClick.bind(this);
    (this: any)._onIntervalMarkerSelect = this._onIntervalMarkerSelect.bind(this);
  }

  _onZoomButtonClick(start: Milliseconds, end: Milliseconds) {
    const { addRangeFilterAndUnsetSelection, zeroAt } = this.props;
    addRangeFilterAndUnsetSelection(start - zeroAt, end - zeroAt);
  }

  _onIntervalMarkerSelect(threadIndex: ThreadIndex, start: Milliseconds, end: Milliseconds) {
    const { timeRange, updateProfileSelection, changeSelectedThread } = this.props;
    updateProfileSelection({
      hasSelection: true,
      isModifying: false,
      selectionStart: Math.max(timeRange.start, start),
      selectionEnd: Math.min(timeRange.end, end),
    });
    changeSelectedThread(threadIndex);
  }

  render() {
    const {
      profile, className, threadOrder, changeThreadOrder, selection,
      updateProfileSelection, timeRange, zeroAt,
    } = this.props;
    const threads = profile.threads;

    return <TimeSelectionScrubber className={`${className}Header`}
                           zeroAt={zeroAt}
                           rangeStart={timeRange.start}
                           rangeEnd={timeRange.end}
                           minSelectionStartWidth={profile.meta.interval}
                           selection={selection}
                           onSelectionChange={updateProfileSelection}
                           onZoomButtonClick={this._onZoomButtonClick}>
      <div className={`${className}HeaderIntervalMarkerOverviewContainer ${className}HeaderIntervalMarkerOverviewContainerJank`}>
        {
          threadOrder.map(threadIndex => {
            const threadName = threads[threadIndex].name;
            const processType = threads[threadIndex].processType;
            return (
              ((threadName === 'GeckoMain' && processType !== 'plugin') ?
                <ProfileThreadJankOverview className={`${className}HeaderIntervalMarkerOverview ${className}HeaderIntervalMarkerOverviewJank`}
                                           rangeStart={timeRange.start}
                                           rangeEnd={timeRange.end}
                                           threadIndex={threadIndex}
                                           key={threadIndex}
                                           onSelect={this._onIntervalMarkerSelect}
                                           isModifyingSelection={selection.isModifying} /> : null)
            );
          })
        }
      </div>
      <div className={`${className}HeaderIntervalMarkerOverviewContainer ${className}HeaderIntervalMarkerOverviewContainerGfx`}>
        {
          threadOrder.map(threadIndex => {
            const threadName = threads[threadIndex].name;
            const processType = threads[threadIndex].processType;
            return (
              (((threadName === 'GeckoMain' || threadName === 'Compositor' || threadName ==='Renderer') && processType !== 'plugin') ?
                <ProfileThreadTracingMarkerOverview className={`${className}HeaderIntervalMarkerOverview ${className}HeaderIntervalMarkerOverviewGfx ${className}HeaderIntervalMarkerOverviewThread${threadName}`}
                                                    rangeStart={timeRange.start}
                                                    rangeEnd={timeRange.end}
                                                    threadIndex={threadIndex}
                                                    key={threadIndex}
                                                    onSelect={this._onIntervalMarkerSelect}
                                                    isModifyingSelection={selection.isModifying} /> : null)
            );
          })
        }
      </div>
      <OverflowEdgeIndicator className={`${className}HeaderOverflowEdgeIndicator`}>
        {<Reorderable tagName='ol'
                     className={`${className}HeaderThreadList`}
                     order={threadOrder}
                     orient='vertical'
                     onChangeOrder={changeThreadOrder}>
          {
            threads.map((thread, threadIndex) =>
              <ProfileThreadHeaderBar key={threadIndex}
                                      index={threadIndex}
                                      interval={profile.meta.interval}
                                      rangeStart={timeRange.start}
                                      rangeEnd={timeRange.end}/>
            )
          }
        </Reorderable>}
      </OverflowEdgeIndicator>
    </TimeSelectionScrubber>;
  }
}

ProfileViewerHeader.propTypes = {
  profile: PropTypes.object.isRequired,
  className: PropTypes.string.isRequired,
  threadOrder: PropTypes.array.isRequired,
  changeThreadOrder: PropTypes.func.isRequired,
  selection: PropTypes.object.isRequired,
  updateProfileSelection: PropTypes.func.isRequired,
  addRangeFilterAndUnsetSelection: PropTypes.func.isRequired,
  timeRange: PropTypes.object.isRequired,
  zeroAt: PropTypes.number.isRequired,
  changeSelectedThread: PropTypes.func.isRequired,
};

export default connect(state => ({
  profile: getProfile(state),
  selection: getProfileViewOptions(state).selection,
  className: 'profileViewer',
  threadOrder: getThreadOrder(state),
  timeRange: getDisplayRange(state),
  zeroAt: getZeroAt(state),
}), actions)(ProfileViewerHeader);
