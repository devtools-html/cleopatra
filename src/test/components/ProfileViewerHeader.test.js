/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import React from 'react';
import ProfileViewerHeader from '../../components/header/ProfileViewerHeader';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { storeWithProfile } from '../fixtures/stores';
import { getProfileForUnfilteredCallTree } from '../fixtures/profiles/profiles-for-call-trees';
import mockCanvasContext from '../fixtures/mocks/canvas-context';
import MockedReactDOM from 'react-dom';

jest.useFakeTimers();

describe('calltree/ProfileCallTreeView', function() {
  (MockedReactDOM: any).__findDOMNode = function(component) {
    // Mock out the findDOMNode call for WithSize.
    if (
      component.props &&
      component.props.className === 'profileViewerHeader'
    ) {
      return { getBoundingClientRect: () => _getBoundingBox(300, 300) };
    } else {
      throw new Error(
        'Another findDOMNode call needs to be mocked out for this test.'
      );
    }
  };

  it('renders the header', () => {
    window.requestAnimationFrame = fn => setTimeout(fn, 0);
    window.devicePixelRatio = 1;
    const ctx = mockCanvasContext();
    /**
     * Mock out any created refs for the components with relevant information.
     */
    function createNodeMock(element) {
      // <TimelineCanvas><canvas /></TimelineCanvas>
      if (element.type === 'canvas') {
        return {
          getBoundingClientRect: () => _getBoundingBox(200, 300),
          getContext: () => ctx,
          style: {},
        };
      }
      return null;
    }

    const header = renderer.create(
      <Provider store={storeWithProfile(getProfileForUnfilteredCallTree())}>
        <ProfileViewerHeader />
      </Provider>,
      { createNodeMock }
    );

    // Flush any requestAnimationFrames.
    jest.runAllTimers();

    const drawCalls = ctx.__flushDrawLog();

    expect(header.toJSON()).toMatchSnapshot();
    expect(drawCalls).toMatchSnapshot();

    delete window.requestAnimationFrame;
    delete window.devicePixelRatio;
  });
});

function _getBoundingBox(width, height) {
  return {
    width,
    height,
    left: 0,
    x: 0,
    top: 0,
    y: 0,
    right: width,
    bottom: height,
  };
}
