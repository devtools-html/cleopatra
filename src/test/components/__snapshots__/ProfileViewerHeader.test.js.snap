// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calltree/ProfileViewerHeader renders the header 1`] = `
<div
  className="profileViewerHeader"
  onMouseDown={[Function]}
  onMouseMove={[Function]}
>
  <div
    className="profileViewerHeaderTimeRuler"
  >
    <ol
      className="timeRulerContainer"
    >
      <li
        className="timeRulerNotch"
        style={
          Object {
            "left": "0px",
          }
        }
      >
        <span
          className="timeRulerNotchText"
        >
          0.000s
        </span>
      </li>
      <li
        className="timeRulerNotch"
        style={
          Object {
            "left": "66.66666666666667px",
          }
        }
      >
        <span
          className="timeRulerNotchText"
        >
          0.002s
        </span>
      </li>
      <li
        className="timeRulerNotch"
        style={
          Object {
            "left": "133.33333333333334px",
          }
        }
      >
        <span
          className="timeRulerNotchText"
        >
          0.004s
        </span>
      </li>
      <li
        className="timeRulerNotch"
        style={
          Object {
            "left": "200px",
          }
        }
      >
        <span
          className="timeRulerNotchText"
        >
          0.006s
        </span>
      </li>
      <li
        className="timeRulerNotch"
        style={
          Object {
            "left": "266.6666666666667px",
          }
        }
      >
        <span
          className="timeRulerNotchText"
        >
          0.008s
        </span>
      </li>
    </ol>
  </div>
  <div
    className="overflowEdgeIndicator profileViewerHeaderOverflowEdgeIndicator"
  >
    <div
      className="overflowEdgeIndicatorEdge topEdge"
    />
    <div
      className="overflowEdgeIndicatorEdge rightEdge"
    />
    <div
      className="overflowEdgeIndicatorEdge bottomEdge"
    />
    <div
      className="overflowEdgeIndicatorEdge leftEdge"
    />
    <div
      className="overflowEdgeIndicatorScrollbox profileViewerHeaderOverflowEdgeIndicatorScrollbox"
      onScroll={[Function]}
    >
      <div
        className="overflowEdgeIndicatorContentsWrapper"
      >
        <ol
          className="profileViewerHeaderThreadList"
          onMouseDown={[Function]}
        >
          <li
            className="profileThreadHeaderBar selected"
            onClick={[Function]}
            style={undefined}
          >
            <div
              className="react-contextmenu-wrapper grippy profileThreadHeaderBarThreadLabel"
              onContextMenu={[Function]}
              onMouseDown={[Function]}
              onMouseOut={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchStart={[Function]}
              title="thread: \\"Main Thread\\" (0)
process: \\"default\\" (0)"
            >
              <h1
                className="profileThreadHeaderBarThreadName"
              >
                Main Thread
              </h1>
            </div>
            <div
              className="profileThreadHeaderBarThreadDetails"
            >
              <div
                className="profileThreadHeaderBarIntervalMarkerOverview profileThreadHeaderBarIntervalMarkerOverviewJank selected"
              >
                <canvas
                  className="profileThreadHeaderBarIntervalMarkerOverviewCanvas profileThreadHeaderBarIntervalMarkerOverviewJankCanvas intervalMarkerTimelineCanvas"
                  onMouseDown={[Function]}
                  onMouseMove={[Function]}
                  onMouseOut={[Function]}
                  onMouseUp={[Function]}
                />
              </div>
              <div
                className="threadStackGraph"
              >
                <canvas
                  className="threadStackGraphCanvas threadStackGraphCanvas"
                  onMouseUp={[Function]}
                />
              </div>
              <div
                className="headerEmptyThreadIndicator"
              />
            </div>
          </li>
          <li
            className="profileThreadHeaderBar"
            onClick={[Function]}
            style={undefined}
          >
            <div
              className="react-contextmenu-wrapper grippy profileThreadHeaderBarThreadLabel"
              onContextMenu={[Function]}
              onMouseDown={[Function]}
              onMouseOut={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchStart={[Function]}
              title="thread: \\"Thread with dropped samples\\" (0)
process: \\"default\\" (0)"
            >
              <h1
                className="profileThreadHeaderBarThreadName"
              >
                Thread with dropped samples
              </h1>
            </div>
            <div
              className="profileThreadHeaderBarThreadDetails"
            >
              <div
                className="profileThreadHeaderBarIntervalMarkerOverview profileThreadHeaderBarIntervalMarkerOverviewJank"
              >
                <canvas
                  className="profileThreadHeaderBarIntervalMarkerOverviewCanvas profileThreadHeaderBarIntervalMarkerOverviewJankCanvas intervalMarkerTimelineCanvas"
                  onMouseDown={[Function]}
                  onMouseMove={[Function]}
                  onMouseOut={[Function]}
                  onMouseUp={[Function]}
                />
              </div>
              <div
                className="threadStackGraph"
              >
                <canvas
                  className="threadStackGraphCanvas threadStackGraphCanvas"
                  onMouseUp={[Function]}
                />
              </div>
              <div
                className="headerEmptyThreadIndicator"
              >
                <div
                  className="headerEmptyThreadIndicatorBlock headerEmptyThreadIndicatorStartup"
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                  style={
                    Object {
                      "left": 0,
                      "width": 66.66666666666667,
                    }
                  }
                />
                <div
                  className="headerEmptyThreadIndicatorBlock headerEmptyThreadIndicatorShutdown"
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                  style={
                    Object {
                      "right": 0,
                      "width": 66.66666666666667,
                    }
                  }
                />
                <div
                  className="headerEmptyThreadIndicatorBlock headerEmptyThreadIndicatorEmptyBuffer"
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                  style={
                    Object {
                      "left": 66.66666666666667,
                      "width": 66.66666666666667,
                    }
                  }
                />
              </div>
            </div>
          </li>
          <li
            className="profileThreadHeaderBar"
            onClick={[Function]}
            style={undefined}
          >
            <div
              className="react-contextmenu-wrapper grippy profileThreadHeaderBarThreadLabel"
              onContextMenu={[Function]}
              onMouseDown={[Function]}
              onMouseOut={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchStart={[Function]}
              title="thread: \\"Thread with dropped samples\\" (0)
process: \\"default\\" (0)"
            >
              <h1
                className="profileThreadHeaderBarThreadName"
              >
                Thread with dropped samples
              </h1>
            </div>
            <div
              className="profileThreadHeaderBarThreadDetails"
            >
              <div
                className="profileThreadHeaderBarIntervalMarkerOverview profileThreadHeaderBarIntervalMarkerOverviewJank"
              >
                <canvas
                  className="profileThreadHeaderBarIntervalMarkerOverviewCanvas profileThreadHeaderBarIntervalMarkerOverviewJankCanvas intervalMarkerTimelineCanvas"
                  onMouseDown={[Function]}
                  onMouseMove={[Function]}
                  onMouseOut={[Function]}
                  onMouseUp={[Function]}
                />
              </div>
              <div
                className="threadStackGraph"
              >
                <canvas
                  className="threadStackGraphCanvas threadStackGraphCanvas"
                  onMouseUp={[Function]}
                />
              </div>
              <div
                className="headerEmptyThreadIndicator"
              >
                <div
                  className="headerEmptyThreadIndicatorBlock headerEmptyThreadIndicatorStartup"
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                  style={
                    Object {
                      "left": 0,
                      "width": 66.66666666666667,
                    }
                  }
                />
                <div
                  className="headerEmptyThreadIndicatorBlock headerEmptyThreadIndicatorShutdown"
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                  style={
                    Object {
                      "right": 0,
                      "width": 66.66666666666667,
                    }
                  }
                />
                <div
                  className="headerEmptyThreadIndicatorBlock headerEmptyThreadIndicatorEmptyBuffer"
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                  style={
                    Object {
                      "left": 66.66666666666667,
                      "width": 66.66666666666667,
                    }
                  }
                />
              </div>
            </div>
          </li>
        </ol>
      </div>
    </div>
  </div>
  <div
    className="timeSelectionScrubberHoverIndicator"
    style={
      Object {
        "left": "0",
        "visibility": "hidden",
      }
    }
  />
</div>
`;

exports[`calltree/ProfileViewerHeader renders the header 2`] = `
Array [
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "fillRect",
    0,
    NaN,
    22.22222222222222,
    NaN,
  ],
  Array [
    "fillRect",
    22.22222222222222,
    NaN,
    22.22222222222222,
    NaN,
  ],
  Array [
    "fillRect",
    44.44444444444444,
    NaN,
    22.22222222222222,
    NaN,
  ],
  Array [
    "fillRect",
    66.66666666666666,
    NaN,
    22.22222222222222,
    NaN,
  ],
  Array [
    "fillRect",
    88.88888888888889,
    NaN,
    22.22222222222222,
    NaN,
  ],
  Array [
    "fillRect",
    111.11111111111111,
    NaN,
    22.22222222222222,
    NaN,
  ],
  Array [
    "fillRect",
    133.33333333333331,
    NaN,
    22.22222222222222,
    NaN,
  ],
  Array [
    "fillRect",
    155.55555555555554,
    NaN,
    22.22222222222222,
    NaN,
  ],
  Array [
    "fillRect",
    177.77777777777777,
    NaN,
    22.22222222222222,
    NaN,
  ],
  Array [
    "set fillStyle",
    "#003eaa",
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "fillRect",
    88.88888888888889,
    0,
    22.22222222222222,
    300,
  ],
  Array [
    "fillRect",
    111.11111111111111,
    0,
    22.22222222222222,
    300,
  ],
  Array [
    "fillRect",
    133.33333333333331,
    75,
    22.22222222222222,
    225,
  ],
  Array [
    "set fillStyle",
    "#003eaa",
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "fillRect",
    88.88888888888889,
    0,
    22.22222222222222,
    300,
  ],
  Array [
    "fillRect",
    111.11111111111111,
    0,
    22.22222222222222,
    300,
  ],
  Array [
    "fillRect",
    133.33333333333331,
    75,
    22.22222222222222,
    225,
  ],
  Array [
    "set fillStyle",
    "#003eaa",
  ],
  Array [
    "clearRect",
    0,
    0,
    300,
    300,
  ],
  Array [
    "scale",
    1,
    1,
  ],
  Array [
    "scale",
    1,
    1,
  ],
  Array [
    "clearRect",
    0,
    0,
    300,
    300,
  ],
  Array [
    "scale",
    1,
    1,
  ],
  Array [
    "scale",
    1,
    1,
  ],
  Array [
    "clearRect",
    0,
    0,
    300,
    300,
  ],
  Array [
    "scale",
    1,
    1,
  ],
  Array [
    "scale",
    1,
    1,
  ],
]
`;
